src/representations/calculator_spherical_expansion.hh

```diff
@@ -321,18 +336,73 @@ namespace rascal {
         this->hyp1f1_calculator.precompute(this->max_radial, this->max_angular);
       }
 
+      // this anymore or remove the unncessary template parameter
+      /* Define the contribution from a neighbour atom to the expansion
+       * without requiring a cluster object os it can be used for benchmarks.
+       */
+      template <AtomicSmearingType AST>
+      Matrix_t compute_contribution(double distance,
```

Comment:
No, I just removed them ^^ :smile: 

More seriously, why would you use `const` when passing by value? It has no effect.

Reply:
So the compiler can potentially optimize (and can be placed in read only memory) and more importantly: it is ensured that the value is not changed within the function (or reused) in subsequent changes to the code, when it is not intended to by the author?

Also to make clear to the reader that it is so, when reading the function?

What do you mean by _no effect_? In that it does not change the calling variable, yes of course, but see there is more to it, like written above.

Reply:
> So the compiler can potentially optimize (and can be placed in read only memory) 

Do you have more information on this? I have never heard of using const variables for optimization, especially related to placing data in read-only memory. Except for global/static variables, they should live on the stack, which is always read-write.

---

> What do you mean by no effect? In that it does not change the calling variable, yes of course, but see there is more to it, like written above.

Mostly that is has no effect on the function ABI or API: the presence of const does not change which variables can be passed to the function.

As you said, it has some effect (that I overlooked, sorry) on the function implementation  --- see below.

>  it is ensured that the value is not changed within the function (or reused) in subsequent changes to the code, when it is not intended to by the author?

This is only true if the const is applied at the function definition, since the const has no effect at the function declaration.

```cpp
#include <iostream>
static void bar(const double a);
int main() {
    bar(3);
    return 0;
}
void bar(double a) {
    a += 3;
    std::cout << a << std::endl;
}
```

I don't really see the gain by doing this for non-pointer primitives, but I can agree to disagree on this!

Reply:
Ok, if I catch this somewhere in the code, I usually do not go to the definition/declaration to add another comment for a suggested const change. I would say this is implied.

And for the optimization, probably not for cases of objects, but primitive types can be used e.g. as directly incorporated into machine code, eliminate code if it is in an expression, loop unrolling.
I am actually not sure about the actual optimization possibilities. It is something I heard/learned years ago in a lecture and assumed to be valid. And looking at it with a quick search reveals that it seems to be bogus, see e.g. [here](http://www.gotw.ca/gotw/081.htm). Bottom line of this article seems 
>Avoid passing objects by const value. Prefer passing them by const reference instead, unless they're cheap-to-copy objects like ints. 
It's a common belief that const-correctness helps compilers generate tighter code. Const is indeed a Good Thing, but the point of this issue of GotW is that const is mainly for humans, rather than for compilers and optimizers. When it comes to writing safe code, const is a great tool that lets programmers write safer code with compiler checking and enforcement. Even when it comes to optimization, const is still principally useful as a tool that lets human class designers better implement handcrafted optimizations, and less so as a tag for omniscient compilers to automatically generate better code.

But my main point is the other one you also cited: using const as an indicator from the author that this variable is not changed (and should not be changed). And not be reused. At one point in the interpolator function a double is reused and I find it potentially misleading to reuse variables. Especially if they are primitive types. Just makes it much less clear. And since we mostly read code...

Bottom line: learned something :)

Reply:
> And for the optimization, probably not for cases of objects, but primitive types can be used e.g. as directly incorporated into machine code, eliminate code if it is in an expression, loop unrolling.

This is what constexpr is about in my understanding: the value must be known at compile-time to allow all these optimization.

Reply:
Yes... I totally see your point and it makes total sense. So I guess this is resolved in
- `const` is not more effiecient, but:
- use `const` in cases like this if it is not changed to signal to reader/later developer that it is not to be reused/changed within the function?